src/state.py

```diff
@@ -0,0 +1,119 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""The module for managing charm state."""
+
+import logging
+import os
+import typing
+from dataclasses import dataclass
+
+import ops
+from pydantic import BaseModel, Field, ValidationError
+
+import metadata
+
+# relation name used for compatibility with machine Jenkins server charm.
+SLAVE_RELATION = "slave"
+# agent relation name
+AGENT_RELATION = "agent"
+
+logger = logging.getLogger()
+
+
+class CharmStateBaseError(Exception):
+    """Represents error with charm state."""
+
+
+class InvalidStateError(CharmStateBaseError):
+    """Exception raised when state configuration is invalid."""
+
+    def __init__(self, msg: str = ""):
+        """Initialize a new instance of the InvalidStateError exception.
+
+        Args:
+            msg: Explanation of the error.
+        """
+        self.msg = msg
+
+
+class JenkinsConfig(BaseModel):
+    """The Jenkins config from juju config values.
+
+    Attrs:
+        server_url: The Jenkins server url.
+        agent_names: Custom Jenkins agent names.
+        agent_tokens: Tokens used to register Jenkins agent to Jenkins server.
+    """
+
+    server_url: str = Field(..., min_length=1)
+    agent_names: typing.List[str] = Field(..., min_items=1)
+    agent_tokens: typing.List[str] = Field(..., min_items=1)
+
+    @classmethod
+    def from_charm_config(cls, config: ops.ConfigData) -> typing.Optional["JenkinsConfig"]:
+        """Instantiate JenkinsConfig from charm config.
+
+        Args:
+            config: Charm configuration data.
+
+        Returns:
+            JenkinsConfig if configuration exists, None otherwise.
+        """
+        server_url = config.get("jenkins_url", None)
+        agent_names = config.get("jenkins_agent_name", None)
+        agent_tokens = config.get("jenkins_agent_token", None)
+        if not server_url and not agent_names and not agent_tokens:
+            return None
```

Comment:
The charm state's attribute `jenkins_config: typing.Optional[JenkinsConfig]` is optional, which denotes that the charm  config method to connect to Jenkins server is not set, hence the connection using the relation takes effect. If any of these are given, it is assumed to be intended to be deployed with charm config rather than relations.

Reply:
I'll add a comment to explain this behaviour. Thank you!